### ISSUE TITLE
dynamically added controls & dynamic event handler - "self" access

### ISSUE THREAD

**Issue by RustyJoeM** [gabdube/native-windows-gui#131]
Hello, 

rather clarification question than real issue...

i'm trying to build a Frame with dynamically added/removed ComboBoxes, and need to resolve / graphically highlight potential duplicates of values set in ComboBoxes.

Two approaches seem feasible:

- one "static" event handler via derive annotations - when i try to add

```
#[nwg_events(OnComboxBoxSelection: [FillerApp::recolor_dropdowns(SELF)])]
tokens_frame: nwg::Frame,
```

and later add ComboBoxes, event handler are not triggered - not sure of this is expectable or not, due to comments in the  guide that most of events should be "forwarded" to parent controls in a way?

Other approach - dynamically adding event handler together with comboboxes, attaching it to the frame. This seems to work, and i start getting event triggers as expected. However, i run into (obviously) lifetime errors, due to my need to use "self" in the handler's closure to get all the dropdown current values, and optionally change some of the dropdown's colors...

Is there some recommended pattern for such scenario?
(getting access to "self" / self.members in dynamically handled NativeUI children/methods)

i found notice of RC_SELF in regard to derive macros, but it's not used in any example, and derive event handler seems not to work for my case either way.

thank you!

**Comment by gabdube** [comment]
For your first solution, one static event handler, this is a limitation of the derive macro. `nwg_events` will only trigger the callback for the exact control event. Because a Frame won't trigger a `OnComboxBoxSelection` itself, the event is always ignored.

For you other approach, it's how it should work.  Working on an example, I found out there was an issue with RC_SELF in the derive macro. It's already fixed, just need to push it to the repo.

Here's how I would solve your problem. It uses one callback created `OnInit` that reference the application data (using RC_SELF).


```rust
extern crate native_windows_derive as nwd;
extern crate native_windows_gui as nwg;

use std::cell::RefCell;
use std::rc::Rc;

use nwg::NativeUi;
use nwd::NwgUi;

use nwg::stretch::{geometry::{Size, Rect},style::{Dimension, FlexDirection, JustifyContent, Style},};

const PT_10: Dimension = Dimension::Points(10.0);
const PAD: Rect<Dimension> = Rect { start: PT_10, end: PT_10, top: PT_10, bottom: PT_10 };

#[derive(Default, NwgUi)]
pub struct MyApp {

    #[nwg_control(title: "test", size: (400, 300), position: (80, 60))]
    #[nwg_events(OnWindowClose: [nwg::stop_thread_dispatch()], OnInit: [MyApp::setup_combo_callback(RC_SELF)])]
    window: nwg::Window,
    
    #[nwg_layout(parent: window, padding: PAD, auto_spacing: None, flex_direction: FlexDirection::Column, justify_content: JustifyContent::Center)]
    main_layout: nwg::FlexboxLayout,

    #[nwg_control(text: "Add combobox")]
    #[nwg_layout_item(layout: main_layout, size: Size { width: Dimension::Percent(1.0), height: Dimension::Points(40.0)})]
    #[nwg_events(OnButtonClick: [MyApp::add_combobox])]
    button: nwg::Button,

    #[nwg_control(parent: window, flags: "VISIBLE")]
    #[nwg_layout_item(layout: main_layout, size: Size { width: Dimension::Percent(1.0), height: Dimension::Points(290.0)})]
    frame: nwg::Frame,

    #[nwg_layout(parent: frame)]
    frame_layout: nwg::FlexboxLayout,

    combo_box_handler: RefCell<Option<nwg::EventHandler>>,
    combo_boxes: RefCell<Vec<nwg::ComboBox<String>>>,
}

impl MyApp {

    fn setup_combo_callback(app_rc: &Rc<MyApp>) {
        // Clone the app and store it in our own callback
        let app = app_rc.clone();
        let handler = nwg::bind_event_handler(&app_rc.frame.handle, &app_rc.window.handle, move |evt, _evt_data, handle| {
            match evt {
                nwg::Event::OnComboxBoxSelection => { 

                    // Fetch the combobox here
                    let boxes = app.combo_boxes.borrow();
                    for (index, combo) in boxes.iter().enumerate() {
                        if combo.handle == handle {
                            println!("Accessed combobox #{}", index);
                        }
                    }

                },
                _ => {}
            }
        });

        *app_rc.combo_box_handler.borrow_mut() = Some(handler);
    }

    fn add_combobox(&self) {
        let mut new_dd: nwg::ComboBox<String> = nwg::ComboBox::<String>::default();
        let coll = vec![String::from("one"), String::from("two"), String::from("three")];
        nwg::ComboBox::builder()
            .collection(coll)
            .parent(&self.frame)
            .build(&mut new_dd)
            .expect("Failed to create token");

        let style = Style {
            size: Size { width: Dimension::Percent(1.0), height: Dimension::Points(40.0) },
            ..Default::default()
        };

        self.frame_layout
            .add_child(&new_dd, style)
            .expect("Failed to add token to layout");

        self.combo_boxes.borrow_mut().push(new_dd);
    }
}

fn main() {
    nwg::init().expect("Failed to init Native Windows GUI");
    let _app = MyApp::build_ui(Default::default()).expect("Failed to build UI");
    nwg::dispatch_thread_events();
}

```

**Comment by RustyJoeM** [comment]
Thanks again for awesome feedback :) coming from lazy ansi-c / python / javascript world, the Rust's approaches sometimes are a bit hard to process... i was imaging intricate combinations of refcells/lifetime declarations that are out of my grasp for now...

Cloning the ui struct in OnInit feels like neat simple approach that did not occur to me - of course it can have some disadvantages in specific scenarios, but can be easily dismissed for my app structure/scale.

As a side comment, cudos on the lib/crate as a whole, currently feels to be best Option<> (pun intended) for GUI if windows target is sufficient. No external run-time deps - code completion/hinting etc. works rather nicely in vscode with rust-analyzer, many examples etc.

(not sure if you want to keep this opened because of the RC_SELF issue you mention, or can close (my query for help is imho fulfilled :) )

**Comment by RustyJoeM** [comment]
erm, seems that i may need to enable some extra features or wait for the menitoned fix...

```
proc-macro derive panicked
message: Unknown callback argument: RC_SELF. Should be one of those values: ["SELF", "CTRL", "HANDLE", "EVT", "EVT_DATA", "EVT_UI"]rustc
```

**Comment by gabdube** [comment]
If you pull master, it should work now. It's has been published to `native_windows_derive 1.0.2`. If it works now, you can close this issue.
 
And thanks for the kind words! I know using `Refcell` is kind of rough. That's the limitation of using win32 without any underlying rust state.

**Comment by RustyJoeM** [comment]
thank you, with 1.0.2 it works like a charm!
